File: service_worker/service_worker.js
// service_worker.js — TactixCore Reborn Service Worker

const CACHE_NAME = 'tactixcore-cache-v1';
const ASSETS_TO_CACHE = [
  '/',
  '/index.html',
  '/style.css',
  '/script.js',
  '/config/config.ini',
  '/config/config.py',
  '/offline.html'  // Optional offline fallback page, create if needed
];

// Install event — cache core assets
self.addEventListener('install', (event) => {
  console.log('[ServiceWorker] Installing and caching assets');
  event.waitUntil(
    caches.open(CACHE_NAME).then((cache) => {
      return cache.addAll(ASSETS_TO_CACHE);
    }).then(() => self.skipWaiting())
  );
});

// Activate event — clean old caches
self.addEventListener('activate', (event) => {
  console.log('[ServiceWorker] Activating and cleaning old caches');
  event.waitUntil(
    caches.keys().then((cacheNames) => {
      return Promise.all(
        cacheNames.map((name) => {
          if (name !== CACHE_NAME) {
            console.log('[ServiceWorker] Deleting old cache:', name);
            return caches.delete(name);
          }
        })
      );
    }).then(() => self.clients.claim())
  );
});

// Fetch event — respond with cache first, fallback to network
self.addEventListener('fetch', (event) => {
  event.respondWith(
    caches.match(event.request).then((cachedResponse) => {
      if (cachedResponse) {
        // Serve from cache
        return cachedResponse;
      }
      // Fetch from network and cache dynamically
      return fetch(event.request).then((networkResponse) => {
        // Check valid response
        if (!networkResponse || networkResponse.status !== 200 || networkResponse.type !== 'basic') {
          return networkResponse;
        }
        // Clone response to cache it
        const responseToCache = networkResponse.clone();
        caches.open(CACHE_NAME).then((cache) => {
          cache.put(event.request, responseToCache);
        });
        return networkResponse;
      }).catch(() => {
        // If offline and request fails, show offline fallback for navigations
        if (event.request.mode === 'navigate') {
          return caches.match('/offline.html');
        }
      });
    })
  );
});

File: service_worker/service_worker.py
# service_worker.py
import os
import time

CACHE = {}
CACHE_NAME = "tactixcore-cache-v1"

def cache_file(file_path):
    if os.path.exists(file_path):
        with open(file_path, 'rb') as f:
            CACHE[file_path] = f.read()
        print(f"Cached: {file_path}")
    else:
        print(f"File not found, cannot cache: {file_path}")

def cache_static_assets():
    assets = [
        'index.html',
        'script.js',
        'style.css',
        'config/config.ini',
        'config/config.py',
    ]
    for asset in assets:
        cache_file(asset)

def fetch_from_cache(request):
    if request in CACHE:
        print(f"Serving from cache: {request}")
        return CACHE[request]
    else:
        print(f"Cache miss: {request}, fetching fresh...")
        # In real SW, fetch from network here
        return None

def update_cache():
    print("Updating cache...")
    cache_static_assets()
    print("Cache updated.")

if __name__ == "__main__":
    print("Service Worker (simulated) starting...")
    update_cache()
    # Simulate fetch event
    fetch_from_cache('index.html')
    time.sleep(2)
    fetch_from_cache('unknown.file')
